### PR TITLE
[training migration] Add ProfilingConfig dataclass

### PR DESCRIPTION
**Pull request by maanug-nv**
# What does this PR do ?

1. Create dataclass in MLM
2. Get MLM arguments to use dataclass

:warning: For major changes (either in lines of code or in its impact), please make sure to first share discuss a design-doc with the team.  

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

**Comment by maanug-nv** [comment]
depends on #2306

File: megatron/training/config.py
Comment (lmcafee-nvidia) [inline megatron/training/config.py:1]: we removed 7 args from `arguments.py` and are adding 9 args into the config ... it looks like `use_nsys_profiler` and `nvtx_ranges` are new. What are these two args used for?

Comment (maanug-nv) [reply megatron/training/config.py:1]: `use_nsys_profiler` will generate `--profile`. It seems like currently in MLM `args.profile` enables Nsys profiling and is also needed to enable torch profiler. This is disambiguated in [Megatron Bridge](https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/main/src/megatron/bridge/training/profiling.py), we control each with `use_nsys_profiler` and `use_pytorch_profiler` respectively, but not upstreaming that in this PR.

`record_shapes` and `nvtx_ranges` are new. i'm [excluding](https://github.com/NVIDIA/Megatron-LM/pull/2393/changes#diff-42fa19ec8893eabf951a5bb21edfc0dfe7c9c8949d5087ab133f5897ea0e3213R2068) those from arg generation for now. In Megatron Bridge, the former is just passed to [torch profiler](https://docs.pytorch.org/docs/stable/profiler.html). Latter will call [this](https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/utils.py#L2066).

Comment (Wohox) [reply megatron/training/config.py:1]: 2 Questions:
- Wondering is it possible to use decorators to exclude the arguments?
- Do you have plans to remove global vars for profiler as well? (In the roadmap?)

Comment (maanug-nv) [reply megatron/training/config.py:1]: I doubt it's possible to exclude specific attributes using a decorator, I'm not sure how you would even apply a decorator to a dataclass attribute. But I could add support for some flag in the attribute metadata to indicate exclusion, e.g. `field(..., metadata={"argparse_meta": {"exclude": True})`.

We won't remove any global vars in core, but we have stopped using global vars in Megatron Bridge training loop, and plan to upstream that change. We have [this](https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/main/src/megatron/bridge/training/state.py#L108) state managament object instead. Although, I don't see any global vars related to profiling in megatron.training , which global vars are you referring to?

**Review by lmcafee-nvidia — APPROVED** [review]

**Review by yanring — APPROVED** [review]

**Review by Wohox — APPROVED** [review]
LGTM
